src/util/sharing_map.h

```diff
@@ -38,26 +40,46 @@ Author: Daniel Poetzl
 
 // clang-format off
 #define SHARING_MAPT(R) \
-  template <class keyT, class valueT, class hashT, class equalT> \
-  R sharing_mapt<keyT, valueT, hashT, equalT>
+  template < \
+    typename keyT, \
+    typename valueT, \
+    bool fail_if_equal, \
+    typename hashT, \
+    typename equalT> \
+  R sharing_mapt<keyT, valueT, fail_if_equal, hashT, equalT>
 
 #define SHARING_MAPT2(CV, ST) \
-  template <class keyT, class valueT, class hashT, class equalT> \
-  CV typename sharing_mapt<keyT, valueT, hashT, equalT>::ST \
-    sharing_mapt<keyT, valueT, hashT, equalT>
+  template < \
+    typename keyT, \
+    typename valueT, \
+    bool fail_if_equal, \
+    typename hashT, \
+    typename equalT> \
+  CV typename sharing_mapt<keyT, valueT, fail_if_equal, hashT, equalT>::ST \
+    sharing_mapt<keyT, valueT, fail_if_equal, hashT, equalT>
 
 #define SHARING_MAPT3(T, CV, ST) \
```

Comment:
This macros could use better names, Which one do I want, and why?

Reply:
Hm I tried that at some point but it turned out to be difficult to distill the differences between the macros into meaningful names. I'd suggest keeping the current names and instead add some documentation to them.

Reply:
Maybe then we shouldn't have that macro in the first place if it isn't possible to name its parameters in a meaningful way? Then really the information exclusively lies with the points where the macro is used. In that case I'd suggest to just expand the macro at all uses, at the expense of some code duplication.

Reply:
I can give the macro parameters better names and document the parameters and the macros. Just giving each macro a name that describes what it does and at the same time makes the distinctions between them clear seems hard. I would prefer though having the macros just to abbreviate those template expressions as otherwise the definitions of the methods and static variables become kind of cluttered.

Reply:
I'm happy with whatever makes this code easier to use. The sharing maps are really useful and efficient, but they are very hard to use, because their API is very easy to read.

Reply:
I agree, the sharing map is somewhat hard to use. My goal is that with this PR it will get easier. The macros above only appear in the definitions btw so don't show up in the interface declarations. Also I'll move the documentation of the public methods from the definitions to the declarations.